suite/tests/client-interface/drbbdup-emul-reg-clobber-test.dll.c

```diff
@@ -0,0 +1,311 @@
+                             XINST_CREATE_cmp(drcontext, opnd_create_reg(DR_REG_START_32),
+                                              OPND_CREATE_INT32(-1)));
+}
+
+static dr_emit_flags_t
+instrument_instr(void *drcontext, void *tag, instrlist_t *bb, instr_t *instr,
+                 instr_t *where, bool for_trace, bool translating, uintptr_t encoding,
+                 void *user_data, void *orig_analysis_data, void *analysis_data)
+{
+    bool is_first, is_last;
+    drbbdup_status_t res;
+    per_block_t *per_block = (per_block_t *)analysis_data;
+
+    res = drbbdup_is_first_instr(drcontext, instr, &is_first);
+    CHECK(res == DRBBDUP_SUCCESS, "failed to check whether instr is first");
+    if (is_first) {
+        per_block->saw_first = true;
+        /* Ensure DR_EMULATE_REST_OF_BLOCK didn't leak through. */
+        const emulated_instr_t *emul_info;
+        if (drmgr_in_emulation_region(drcontext, &emul_info)) {
+            /* It might be the rep string, which sets DR_EMULATE_INSTR_ONLY.
+             * XXX: We could try to pass a flag: but drbbdup won't let
+             * us pass from app2app; we'd need TLS or a global.
+             */
+            CHECK(!TEST(DR_EMULATE_REST_OF_BLOCK, emul_info->flags) ||
+                      TEST(DR_EMULATE_INSTR_ONLY, emul_info->flags),
+                  "DR_EMULATE_REST_OF_BLOCK leaked through!");
+        }
+    }
+
+#ifdef X86
+    /* Cause drutil's expanded "rep movs" to crash if drbbdup sets
+     * DRREG_IGNORE_CONTROL_FLOW (i#5398).  We do not spill rcx until after the
+     * rcx==0 path jumps over the OP_movs, such that drreg's restore before the
+     * OP_loop writes the wrong value.
+     */
+    if (instr_get_opcode(instr) == OP_movs) {
+        saw_movs = true;
+        reg_id_t reg_clobber;
+        drvector_t rvec;
+        drreg_init_and_fill_vector(&rvec, false);
+        drreg_set_vector_entry(&rvec, DR_REG_XCX, true);
+        if (drreg_reserve_register(drcontext, bb, where, &rvec, &reg_clobber) !=
+            DRREG_SUCCESS)
+            CHECK(false, "failed to reserve scratch register\n");
+        drvector_delete(&rvec);
+        instrlist_meta_preinsert(bb, where,
+                                 XINST_CREATE_load_int(drcontext,
+                                                       opnd_create_reg(reg_clobber),
+                                                       OPND_CREATE_INT32(-1)));
+        if (drreg_reserve_aflags(drcontext, bb, instr) != DRREG_SUCCESS)
+            CHECK(false, "cannot reserve aflags");
+        write_aflags(drcontext, bb, instr);
+        if (drreg_unreserve_aflags(drcontext, bb, instr) != DRREG_SUCCESS)
+            CHECK(false, "cannot unreserve aflags");
+        if (drreg_unreserve_register(drcontext, bb, where, reg_clobber) != DRREG_SUCCESS)
+            CHECK(false, "failed to unreserve scratch register\n");
+    }
+#endif
+
+    /* Ensure the drmgr emulation API works for a final special instr where "instr" !=
```

Comment:
This again can all be removed (the entire rest of this function, right?) instead of being a redundant copy from the other test which makes this test harder to understand.

Reply:
Removed.